<a href="https://colab.research.google.com/github/aerosense-ai/notebooks/blob/main/process_juju.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/aerosense-ai/notebooks/blob/main/pre_process.ipynb)

In [2]:
!python --version

Python 3.8.15


# Python 3.9
If we want we can force Python 3.9 or even 3.11

**IMPORTANT**: Runtime-> Restart Runtime after changes!

In [2]:
#install python 3.9
!sudo apt-get update -y
!sudo apt-get install python3.9 python3.9-dev python3.9-distutils libpython3.9-dev

#change alternatives
!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.7 1
!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 2

# install pip
!sudo apt-get install python3-pip
!python -m pip install --upgrade pip

#install colab's dependencies
!python3 -m pip install ipython ipython_genutils ipykernel jupyter_console prompt_toolkit httplib2 astor

#some more depedences
!python3 -m pip install --upgrade six nbformat

# link to the old google package
!ln -s /usr/local/lib/python3.7/dist-packages/google \
       /usr/local/lib/python3.9/dist-packages/google

#check python version
!python --version

Get:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Hit:2 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:3 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:4 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Get:5 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [83.3 kB]
Ign:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Get:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease [1,581 B]
Hit:8 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Get:9 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]
Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Get:11 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease [15.9 kB]
Hit:12 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease
Get:13 http

# Dependences and new code

In [1]:
!pip install --upgrade git+https://github.com/aerosense-ai/aerosense-tools.git@feature/add-pre-process

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/aerosense-ai/aerosense-tools.git (to revision feature/add-pre-process) to /tmp/pip-req-build-zxslay9s
  Running command git clone --filter=blob:none --quiet https://github.com/aerosense-ai/aerosense-tools.git /tmp/pip-req-build-zxslay9s
  Running command git checkout -b feature/add-pre-process --track origin/feature/add-pre-process
  Switched to a new branch 'feature/add-pre-process'
  Branch 'feature/add-pre-process' set up to track remote branch 'feature/add-pre-process' from 'origin'.
  Resolved https://github.com/aerosense-ai/aerosense-tools.git to commit 4f7012cd34fc303c89f9dbb7d16b3736cf612820
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.3/15.3 MB 57.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.0

In [3]:
from google.colab import auth
auth.authenticate_user()

In [4]:
import datetime as dt

from aerosense_tools.queries import BigQuery
from aerosense_tools import plots
from aerosense_tools.preprocess import RawSignal
from aerosense_tools.preprocess import SensorMeasurementSession

client = BigQuery()

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
import pandas as pd
import numpy as np

In [7]:
sensor_types_metadata = {
    "accelerometer": {
        "description": "Accelerometer Measurement Series",
        "variable": "Acceleration [m/s²]",
        "sensors": ["Acc. Direction 1", "Acc. Direction 2", "Acc. Direction 3"],
        },
    "barometer": {
        "description": "Absolute Barometer Measurement Series",
        "variable": "Pressure [Pa]",
        "sensors": ["Abs. Baro. Sensor {}".format(i) for i in range(40)],
    },
    "barometer_thermometer": {
        "description": "Temperature from Absolute Barometer Sensor Measurement Series",
        "variable": "Temperature [C]",
        "sensors": ["Temp. Sensor {}".format(i) for i in range(40)]
    },
    "connection_statistics": {
        "description": "Connection Statistics",
        "variable": [
            "Transmitting Power [W]",
            "Allocated Heap Memory [Bytes]",
            "Received signal strength indication: Raw [dBm]",
            "Received signal strength indication: Filtered [dBm]"
        ],
        "sensors": ["Conn. Stat. {}".format(i) for i in range(4)]
    },
    "differential_barometer": {
        "description": "Differential Barometer Measurement Series",
        "variable": "Differential Pressure [Pa]",
        "sensors": ["Diff. Baro. Sensor {}".format(i) for i in range(5)],
        },
    "gyroscope": {
        "description": "Gyroscope Measurement Series",
        "variable": "Angular Velocity [s⁻¹]",
        "sensors": ["Gyro. Direction 1", "Gyro. Direction 2", "Gyro. Direction 3"],
        },
    "magnetometer": {
        "description": "Magnetometer Measurement Series",
        "variable": "Magnetic Momemnt [A⋅m²]",
        "sensors": ["Mag. Direction 1", "Mag. Direction 2", "Mag. Direction 3"],
        },
}

**WIP** on data processing funcitons we will need in future:

In [ ]:
#WIP    
def correct_height():
    """ One of the several corrections we need to apply on the absolute pressure values
    """

def azimuth_position(imu_orientation):
    """ From accelerometer and gyrometer, we need to know the orientation of the IMU to project the three axes in the axes of the blade. Is it trivial? Do we really need the orientations of the IMU, or can we do without it?
    """

def height_sensors(azimuth,r_pos):
    """ When azimuth position of the blade is known, we can determine the height of the sensors relative to the nacelle height.
    It is certainly possible to interpolate the difference of height without taking into account the actual altitude: Locally, the gradient of pressure might be seen as a linear evolution with the height. (Is it the case on much  bigger wind turbines?)
    """

# Maybe needed much later. Not to develop for now.
def correct_temperature():
    """ If a full set of calibration with pressure and temperature of reference.
    So far it's not required (only for Lyon expe)
        """

# Data overview
Before analysing all the data individually, we need to know what we have a bit more directly and visually

## labbook

we want to know:
- when the data has been recorded: *start_time end_time*
- how good the data is: what type of sensors do we have recorded + points to discard because faulty data (use Yuriy's cleaning process) and make a ratio of discarded points vs complete length, rssi as well?
- what were the wind conditions: wind speed (how?), rotational speed (make a simple algo on IMU to get an idea of rotational speed), average_temperature (average value from the IMU or pressure sensors?)

index | start_time | end_time | baros | IMU | mics | diffP | ratio good data | average blade rotational speed | temperature | zero_val | rssi_mean?
-- | --| -- | -- | --| --|--|-- | -- | -- | -- |--
int? | date | date | boolean | boolean | boolean | boolean | double | double | double | boolean | double

From that we will have an overview to be able to look at relevant data more efficiently.
Questions:
- How to present this?
- Where to store it?
- It should be a table that can be concatenated as soon new measurements are coming, with possibility to add some filtering

## Other parameters to be stored somewhere

- For each sensor node: Blade specs (at one specific location or more general?)
	- shape
	- chord length
- Data about baros
	- baro_index
	- index from LE (+ on suction side, - on pressure side)
	- position along the chord
	- radial position (along the blade)
	- Coeff for pressure and temperature that could be set individually for each of them
- Data about mics
	- baro_index (00 the one in the corner, i0 for those installed streamwise and 0j for those installed spanwise)
	- distance between each sensor from the previous ones (streamwise and  spanwise)
	- their positions along the chord and along r(?)
	- Coeffs for microphones (unique value for all of them might be sufficient)
- Data about IMU
	- position of IMU on the chord
	- orientation of IMU (Matrix of rotation of the data or angle to use complex/quaternion values)

## Some general functions to be added (but not urgent)
1. From temperature and atmospheric pressure estimate:
  - density
  - kinematic and dynamic viscosity

2. Give normal of one point from the blade section:
  - for barometers for example, it's helpful to integrate pressure.
  - need wing shape as reference


In [ ]:
#Get the time serties for barometer by running a SQL query
query_string = f"""
      SELECT datetime
      FROM `greta.sensor_data`
      WHERE sensor_type_reference="barometer" and
      node_id="2" and
      datetime between "2022-07-01 00:00:00.0" and "2022-09-01 00:00:00.0"
      ORDER BY datetime ASC
      """

baros_sample_times=client.query(query_string)

In [ ]:
# Not sure if this can be done more elegantly. 
# And, surely, can be done directly with SQL (I just don't dare).

threshold = dt.timedelta(seconds=60)
baros_gaps = baros_sample_times['datetime'].diff() > threshold

session_starts = baros_sample_times[baros_gaps]
session_ends=baros_sample_times.iloc[session_starts.index-1]

first_session_start = pd.DataFrame({"datetime":[baros_sample_times['datetime'][0]]})
last_session_end = pd.DataFrame({"datetime":[baros_sample_times['datetime'].iat[-1]]})

session_starts=pd.concat([first_session_start, session_starts])
session_ends=pd.concat([session_ends, last_session_end])

sessions = pd.DataFrame({"start_time":session_starts['datetime'].reset_index(drop=True), "end_time":session_ends['datetime'].reset_index(drop=True)})
sessions['session_duration']=sessions["end_time"]-sessions["start_time"]

In [ ]:
sessions

start_time                   end_time  \
0  2022-07-14 11:06:11.715167 2022-07-14 11:06:21.944842   
1  2022-07-21 09:18:33.627776 2022-07-21 09:19:11.496264   
2  2022-07-21 10:39:21.971243 2022-07-21 10:41:55.604312   
3  2022-07-21 10:43:46.818994 2022-07-21 10:44:22.029420   
4  2022-07-21 11:00:03.773041 2022-07-21 11:00:37.015361   
5  2022-07-21 11:04:50.027605 2022-07-21 11:05:03.783538   
6  2022-07-21 13:00:03.960630 2022-07-21 13:00:30.339872   
7  2022-07-21 13:04:54.399324 2022-07-21 13:05:04.039013   
8  2022-07-21 14:44:35.157202 2022-07-21 14:47:03.444745   
9  2022-07-21 15:26:53.133385 2022-07-21 15:28:15.714353   
10 2022-07-21 15:41:15.865465 2022-07-21 15:45:48.773018   
11 2022-07-21 16:49:49.313949 2022-07-21 16:50:05.963454   
12 2022-07-22 08:58:25.855567 2022-07-22 08:58:38.425147   
13 2022-07-22 09:15:43.837768 2022-07-22 09:22:23.672695   
14 2022-07-22 09:24:45.041385 2022-07-22 09:31:06.936444   
15 2022-07-22 09:58:30.369097 2022-07-22 10:01:58.141360   
16 2022-07-22 12:17:36.417765 2022-07-22 12:23:27.383299   
17 2022-07-22 13:01:06.677570 2022-07-22 13:06:06.477508   
18 2022-07-22 14:01:06.561198 2022-07-22 14:06:06.738049   
19 2022-07-22 15:01:06.628489 2022-07-22 15:06:06.628454   
20 2022-07-22 16:01:06.701719 2022-07-22 16:06:06.803389   
21 2022-07-22 17:01:06.763283 2022-07-22 17:06:06.779112   
22 2022-07-22 18:01:06.822573 2022-07-22 18:06:06.906614   
23 2022-07-22 19:01:06.963738 2022-07-22 19:06:06.929668   
24 2022-07-22 20:01:07.099464 2022-07-22 20:06:07.167476   
25 2022-07-22 21:01:07.121778 2022-07-22 21:06:07.206675   
26 2022-07-22 22:01:07.254009 2022-07-22 22:06:07.384462   
27 2022-07-22 23:01:07.369285 2022-07-22 23:06:07.340935   
28 2022-07-23 00:01:07.462523 2022-07-23 00:06:07.528870   
29 2022-07-23 01:01:07.564790 2022-07-23 01:06:07.622577   
30 2022-07-23 02:01:07.655052 2022-07-23 02:06:07.715824   
31 2022-07-23 03:01:07.748110 2022-07-23 03:06:07.829399   
32 2022-07-23 03:19:15.653974 2022-07-23 03:24:51.961514   
33 2022-07-23 03:44:19.647953 2022-07-23 03:45:08.187925   
34 2022-07-23 03:53:24.051323 2022-07-23 03:59:32.014764   
35 2022-07-23 04:01:07.825897 2022-07-23 04:06:07.972167   
36 2022-07-23 05:01:07.913583 2022-07-23 05:06:08.178469   
37 2022-07-23 07:01:08.077070 2022-07-23 07:06:08.252841   
38 2022-07-23 08:43:31.886915 2022-07-23 08:49:07.893172   
39 2022-07-23 10:01:08.344484 2022-07-23 10:06:08.520159   
40 2022-07-23 12:01:08.537584 2022-07-23 12:06:08.628855   
41 2022-07-23 15:01:08.716257 2022-07-23 15:06:08.791163   
42 2022-07-23 16:01:08.851093 2022-07-23 16:02:11.564910   
43 2022-07-30 08:19:59.383374 2022-07-30 08:20:02.023266   
44 2022-07-30 10:07:42.098652 2022-07-30 10:12:22.180432   
45 2022-07-30 11:07:42.164852 2022-07-30 11:12:22.122787   
46 2022-07-30 12:07:42.233502 2022-07-30 12:12:22.273700   
47 2022-07-30 13:07:42.321806 2022-07-30 13:12:22.337018   
48 2022-07-30 14:07:42.428017 2022-07-30 14:12:22.461376   
49 2022-07-30 15:07:42.540008 2022-07-30 15:12:22.550179   
50 2022-07-30 16:07:42.622773 2022-07-30 16:12:22.660772   
51 2022-07-30 17:07:42.733378 2022-07-30 17:12:22.770113   
52 2022-07-30 18:07:42.801240 2022-07-30 18:12:22.852727   
53 2022-07-30 19:07:42.912275 2022-07-30 19:12:22.960261   
54 2022-07-30 20:07:43.010699 2022-07-30 20:12:23.071495   
55 2022-07-30 21:07:43.112581 2022-07-30 21:12:23.133524   
56 2022-07-30 22:07:43.160687 2022-07-30 22:12:23.115859   
57 2022-07-30 23:07:43.244318 2022-07-30 23:12:23.207226   
58 2022-07-31 00:07:43.327652 2022-07-31 00:12:23.373372   

         session_duration  
0  0 days 00:00:10.229675  
1  0 days 00:00:37.868488  
2  0 days 00:02:33.633069  
3  0 days 00:00:35.210426  
4  0 days 00:00:33.242320  
5  0 days 00:00:13.755933  
6  0 days 00:00:26.379242  
7  0 days 00:00:09.639689  
8  0 days 00:02:28.287543  
9  0 days 00:01:22.580968  
10 0 days 00:04:32.907553  
11 0 days 00:00:16.649505  
12 0 days 00:00:12.569580  
13 0 days 00:06:39.834927 

# General notes on the July measurement campain
## Barometers and Barometer-Themometers

### July 21 - July 23
* Unrealistic pressure flactuations of 30kPa are present.
* The signal on some sensors presents abnormal discontinuities: mean pressure can jump by 10kPa
in a discontinious manner.

Ex: 

index | start_time | end_time | duration | note
-- | --| -- | -- | --
8 | 2022-07-21 14:44:35.157202 |	2022-07-21 14:47:03.444745 |	0 days 00:02:28.287543 | Discontinutiy for sensor 10 and 13 at 14:45:45, after which the magnitude of flactuions degreseases significantly. 
27 |	2022-07-22 23:01:07.369285 |	2022-07-22 23:06:07.340935 |	0 days 00:04:59.971650 | Operating, no AoA reversal, no unrealistic fluctions, discontinuities on several sensors
40 |	2022-07-23 12:01:08.537584 |	2022-07-23 12:06:08.628855 |	0 days 00:05:00.091271 | Fluctions are modulated from extremee to reasonable.

### July 30
index | start_time | end_time | duration | note
-- | --| -- | -- | --
44 |	2022-07-30 10:07:42.098652 |	2022-07-30 10:12:22.180432 |	0 days 00:04:40.081780 | Extreme Pressure fluction modulation
46 |	2022-07-30 12:07:42.233502 |	2022-07-30 12:12:22.273700 |	0 days 00:04:40.040198 | Pressure side and suction side switch during rotation

## Turbine operation:

index | start_time | end_time | duration | note
-- | --| -- | -- | --
1|	2022-07-21 09:18:33.627776|	2022-07-21 09:19:11.496264|	0 days 00:00:37.868488| Rotating
...| ...| ...| ... |...
6|	2022-07-21 13:00:03.960630|	2022-07-21 13:00:30.339872|	0 days 00:00:26.379242| Rotating
7|	2022-07-21 13:04:54.399324|	2022-07-21 13:05:04.039013|	0 days 00:00:09.639689| Stopped
8|	2022-07-21 14:44:35.157202|	2022-07-21 14:47:03.444745|	0 days 00:02:28.287543|Rotating
...| ...| ...| ... |...
14| 2022-07-22 09:24:45.041385 |	2022-07-22 09:31:06.936444 |	0 days 00:06:21.895059 | Operating at almost 1 Hz (64 rpm is max for aventa)
16 | 	2022-07-22 12:17:36.417765  | 	2022-07-22 12:23:27.383299  | 	0 days 00:05:50.965534  | Stopping at 12:20
17  | 	2022-07-22 13:01:06.677570  | 	2022-07-22 13:06:06.477508  | 	0 days 00:04:59.799938  |  Starting
18 	 | 2022-07-22 14:01:06.561198  | 	2022-07-22 14:06:06.738049  | 	0 days 00:05:00.176851  | Rotating 0.5-0.6 Hz
19  | 	2022-07-22 15:01:06.628489  | 	2022-07-22 15:06:06.628454  | 	0 days 00:04:59.999965  |  Rotating with high accelerations and vibrations
20  | 	2022-07-22  |  16:01:06.701719  | 	2022-07-22 16:06:06.803389  | 	0 days 00:05:00.101670  | Rotating 0.5-0.6 Hz
...| ...| ...| ... |...
23 |	2022-07-22 19:01:06.963738 |	2022-07-22 19:06:06.929668 |	0 days 00:04:59.965930 | Rotating 0.5-0.6 Hz
24 |	2022-07-22 20:01:07.099464 |	2022-07-22 20:06:07.167476 |	0 days 00:05:00.068012 | Stopped
...| ...| ...| ... |...
26 |	2022-07-22 22:01:07.254009 |	2022-07-22 22:06:07.384462 |	0 days 00:05:00.130453 | Stoppped
27 |	2022-07-22 23:01:07.369285 |	2022-07-22 23:06:07.340935 |	0 days 00:04:59.971650 | Rotating
...| ...| ...| ... |...
31 |	2022-07-23 03:01:07.748110 |	2022-07-23 03:06:07.829399 |	0 days 00:05:00.081289 | Rotating, but w sensor signal losses
...| ...| ...| ... |...
35 |	2022-07-23 04:01:07.825897 |	2022-07-23 04:06:07.972167 |	0 days 00:05:00.146270 | Rotating, but w sensor signal losses
36 |	2022-07-23 05:01:07.913583 |	2022-07-23 05:06:08.178469 |	0 days 00:05:00.264886 | Stopped
37 |	2022-07-23 07:01:08.077070 |	2022-07-23 07:06:08.252841 |	0 days 00:05:00.175771 | Rotating 0.5-0.6 Hz
...| ...| ...| ... |...
40 |	2022-07-23 12:01:08.537584 |	2022-07-23 12:06:08.628855 |	0 days 00:05:00.091271 | Rotating 0.5-0.6 Hz
41 |	2022-07-23 15:01:08.716257 |	2022-07-23 15:06:08.791163 |	0 days 00:05:00.074906 | Stopped
...| ...| ...| ... |...
44 |	2022-07-30 10:07:42.098652 |	2022-07-30 10:12:22.180432 |	0 days 00:04:40.081780 | Rotating 0.5-0.6 Hz
...| ...| ...| ... |...
49 |	2022-07-30 15:07:42.540008 | 	2022-07-30 15:12:22.550179 |	0 days 00:04:40.010171 | Rotating 0.5-0.6 Hz
50 |	2022-07-30 16:07:42.622773 |	2022-07-30 16:12:22.660772 |	0 days 00:04:40.037999 | Starting 
51 |	2022-07-30 17:07:42.733378 |	2022-07-30 17:12:22.770113 |	0 days 00:04:40.036735 | Rotating
52  | 	2022-07-30 18:07:42.801240  | 	2022-07-30 18:12:22.852727  | 	0 days 00:04:40.051487 | Stopped but some blade movement (some wind present)
...| ...| ...| ... |...
56 |	2022-07-30 22:07:43.160687 	| 2022-07-30 22:12:23.115859 |	0 days 00:04:39.955172 | Stopped, used to get P_atm, Not sure if there is wind









# Data Playground
Lets play with Raw data

In [8]:
df, data_limit_applied = client.get_sensor_data(
    installation_reference="aventa-turbine-test",
    node_id="2",
    sensor_type_reference="barometer",
    start=dt.datetime(2022, 7, 21, 10, 39, 00),
    finish=dt.datetime(2022, 7, 21, 10, 42, 00),
    row_limit=10000,
)

In [ ]:
data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
signal_df = df[["datetime"]+data_columns].set_index('datetime')
barometer = RawSignal(signal_df, "barometer")

In [ ]:
barometer.pad_gaps(dt.timedelta(seconds=1))

In [ ]:
barometer.measurement_to_variable()


There seems to be loads of non-sensical values.  Lets remove everything more than 1.5 Atm and less than 0.6  (... that's my arbitrary choice now, before I can come up with something better)

In [ ]:
barometer.dataframe = barometer.dataframe[(barometer.dataframe <= 150e3) & (barometer.dataframe >= 60e3)]

Lets plot just around peak pressure, so that we can look at the data..

In [ ]:
plot_start = dt.datetime(2022, 7, 21, 10, 41, 51, 150000)
plot_finish = dt.datetime(2022, 7, 21, 10, 41, 51, 650000)
time_mask = ((barometer.dataframe.index > plot_start) & (barometer.dataframe.index < plot_finish))
pre_processed_df = barometer.dataframe[time_mask].reset_index()

# Pre-deployment test

In [ ]:
sensor_signals = {}
sensor_sessions={}
sensor_session_times={}

sample_start = dt.datetime(2022, 12, 2)
sample_finish = dt.datetime(2023, 1, 1)

installation = "predep-test-20221130"

sensors_to_query = [
    'barometer',
    'barometer_thermometer',
    'differential_barometer',
    'accelerometer',
    'gyroscope',
    ]

In [ ]:
# Note, it is good to keep the query parameters constant thoughout the day,
# to make use of BQ cache.

for sensor in sensors_to_query:
  df, data_limit_applied = client.get_sensor_data(
      installation_reference=installation,
      node_id="1",
      sensor_type_reference=sensor,
      start=sample_start,
      finish=sample_finish,
      row_limit=1000000000,
  )

  data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
  signal_df = df[["datetime"]+data_columns].set_index('datetime')
  signal_df.columns = sensor_types_metadata[sensor]["sensors"]
  signal=RawSignal(signal_df, sensor)
  signal.pad_gaps(dt.timedelta(seconds=1))
  signal.measurement_to_variable()
  sensor_signals[sensor]=signal
  sensor_sessions[sensor], sensor_session_times[sensor] = signal.extract_measurement_sessions()



In [ ]:
for sensor in sensors_to_query:
  sensor_signals[sensor].dataframe.to_csv("drive/MyDrive/{}_df_dump.csv".format(sensor))

**Loading up data from the drive (while we redesign the BQ)**

In [ ]:
for sensor in sensors_to_query:
  df = pd.read_csv('/content/drive/MyDrive/Share/AeroSense/{}_df_dump.csv'.format(sensor), index_col = 0)
  df.index = pd.to_datetime(df.index, format="%Y-%m-%d %H:%M:%S")
  sensor_signals[sensor] = RawSignal(df, sensor)
  sensor_sessions[sensor], sensor_session_times[sensor] = RawSignal(df, sensor).extract_measurement_sessions()

In [ ]:
sensor_sessions['barometer'][45].dataframe

Abs. Baro. Sensor 0  Abs. Baro. Sensor 1  \
datetime                                                               
2022-12-03 06:41:42.369890         95478.906250         95501.245117   
2022-12-03 06:41:42.379900         95488.842773         95512.622070   
2022-12-03 06:41:42.389909         95496.362305         95509.472656   
2022-12-03 06:41:42.399889         95498.828125         95510.766602   
2022-12-03 06:41:42.409898         95499.560547         95504.321289   
...                                         ...                  ...   
2022-12-03 06:49:42.386903         95546.997070         95571.826172   
2022-12-03 06:49:42.396882         95563.525391         95579.077148   
2022-12-03 06:49:42.406892         95552.099609         95572.387695   
2022-12-03 06:49:42.416901         95547.656250         95561.767578   
2022-12-03 06:49:42.426881         95529.345703         95572.265625   

                            Abs. Baro. Sensor 2  Abs. Baro. Sensor 3  \
datetime                                                               
2022-12-03 06:41:42.369890         95502.612305         95553.417969   
2022-12-03 06:41:42.379900         95517.797852         95551.562500   
2022-12-03 06:41:42.389909         95512.329102         95561.352539   
2022-12-03 06:41:42.399889         95516.577148         95552.392578   
2022-12-03 06:41:42.409898         95510.131836         95551.269531   
...                                         ...                  ...   
2022-12-03 06:49:42.386903         95553.125000         95606.152344   
2022-12-03 06:49:42.396882         95562.548828         95595.239258   
2022-12-03 06:49:42.406892         95556.591797         95596.997070   
2022-12-03 06:49:42.416901         95552.099609         95601.953125   
2022-12-03 06:49:42.426881         95560.693359         95594.921875   

                            Abs. Baro. Sensor 4  Abs. Baro. Sensor 5  \
datetime                                                               
2022-12-03 06:41:42.369890         95474.194336         95496.850586   
2022-12-03 06:41:42.379900         95477.783203         95500.683594   
2022-12-03 06:41:42.389909         95477.880859         95499.755859   
2022-12-03 06:41:42.399889         95475.927734         95494.702148   
2022-12-03 06:41:42.409898         95478.881836         95497.338867   
...                                         ...                  ...   
2022-12-03 06:49:42.386903         95523.950195         95542.919922   
2022-12-03 06:49:42.396882         95517.846680         95543.823242   
2022-12-03 06:49:42.406892         95528.125000         95539.941406   
2022-12-03 06:49:42.416901         95520.214844         95532.934570   
2022-12-03 06:49:42.426881         95508.911133         95534.960938   

                            Abs. Baro. Sensor 6  Abs. Baro. Sensor 7  \
datetime                                                               
2022-12-03 06:41:42.369890         95554.370117         95572.583008   
2022-12-03 06:41:42.379900         95558.276367         95561.376953   
2022-12-03 06:41:42.389909         95560.034180         95573.559570   
2022-12-03 06:41:42.399889         95550.976562         95559.765625   
2022-12-03 06:41:42.409898         95544.653320         95558.471680   
...                                         ...                  ...   
2022-12-03 06:49:42.386903         95604.125977         95590.307617   
2022-12-03 06:49:42.396882         95584.472656         95609.619141   
2022-12-03 06:49:42.406892         95593.359375         95594.873047   
2022-12-03 06:49:42.416901         95589.526367         95592.456055   
2022-12-03 06:49:42.426881         95588.793945         95596.557617   

                            Abs. Baro. Sensor 8  Abs. Baro. Sensor 9  ...  \
datetime                                                              ...   
2022-12-03 06:41:42.369890         95546.801758         95622.851562  ...   
2022-12-03 06:41:42.379900         95561.791992  

In [ ]:
resampled_session=sensor_sessions['barometer'][45].to_constant_timestep(1/100, timeseries_start=dt.datetime(2022,12,3,6,41,42,400000))
resampled_session.dataframe

Abs. Baro. Sensor 0  Abs. Baro. Sensor 1  \
2022-12-03 06:41:42.400         95498.836255         95510.695056   
2022-12-03 06:41:42.410         95499.489475         95504.300663   
2022-12-03 06:41:42.420         95492.505499         95502.357429   
2022-12-03 06:41:42.430         95484.731019         95508.894369   
2022-12-03 06:41:42.440         95488.165069         95492.554011   
...                                      ...                  ...   
2022-12-03 06:49:42.380         95546.289958         95569.233425   
2022-12-03 06:49:42.390         95552.126809         95574.076589   
2022-12-03 06:49:42.400         95559.966300         95576.993407   
2022-12-03 06:49:42.410         95550.719926         95569.090101   
2022-12-03 06:49:42.420         95541.970660         95565.027316   

                         Abs. Baro. Sensor 2  Abs. Baro. Sensor 3  \
2022-12-03 06:41:42.400         95516.505603         95552.380112   
2022-12-03 06:41:42.410         95510.149480         95551.314759   
2022-12-03 06:41:42.420         95511.896938         95555.647685   
2022-12-03 06:41:42.430         95515.267877         95548.673657   
2022-12-03 06:41:42.440         95511.723218         95552.282740   
...                                      ...                  ...   
2022-12-03 06:49:42.380         95556.744743         95609.990955   
2022-12-03 06:49:42.390         95556.049785         95602.765352   
2022-12-03 06:49:42.400         95560.693234         95595.786810   
2022-12-03 06:49:42.410         95555.196952         95598.535948   
2022-12-03 06:49:42.420         95554.768046         95599.769858   

                         Abs. Baro. Sensor 4  Abs. Baro. Sensor 5  \
2022-12-03 06:41:42.400         95475.960526         95494.731417   
2022-12-03 06:41:42.410         95478.834372         95497.355020   
2022-12-03 06:41:42.420         95474.168384         95498.840789   
2022-12-03 06:41:42.430         95468.724610         95489.667184   
2022-12-03 06:41:42.440         95466.458475         95486.950542   
...                                      ...                  ...   
2022-12-03 06:49:42.380         95529.034672         95548.812527   
2022-12-03 06:49:42.390         95522.055905         95543.200277   
2022-12-03 06:49:42.400         95521.048341         95542.614064   
2022-12-03 06:49:42.410         95525.668861         95537.765752   
2022-12-03 06:49:42.420         95516.704940         95533.563776   

                         Abs. Baro. Sensor 6  Abs. Baro. Sensor 7  \
2022-12-03 06:41:42.400         95550.906372         95559.751262   
2022-12-03 06:41:42.410         95544.734580         95558.510446   
2022-12-03 06:41:42.420         95552.515751         95562.348627   
2022-12-03 06:41:42.430         95539.592749         95569.516694   
2022-12-03 06:41:42.440         95547.490506         95553.190981   
...                                      ...                  ...   
2022-12-03 06:49:42.380         95601.701590         95601.402551   
2022-12-03 06:49:42.390         95598.026361         95596.301152   
2022-12-03 06:49:42.400         95587.240838         95605.025784   
2022-12-03 06:49:42.410         95592.169209         95594.122560   
2022-12-03 06:49:42.420         95589.298944         95593.729627   

                         Abs. Baro. Sensor 8  Abs. Baro. Sensor 9  ...  \
2022-12-03 06:41:42.400         95539.098143         95608.462583  ...   
2022-12-03 06:41:42.410         95528.990290         95596.563632  ...   
2022-12-03 06:41:42.420         95506.273420         95575.498167  ...   
2022-12-03 06:41:42.430         95503.382182         95544.908442  ...   
2022-12-03 06:41:42.440         95489.178366         95532.648767  ...   
...                                      ...                  ...  ...   
2022-12-03 06:49:42.380         95466.660323         95511.310504  ...   
2022-12-03 06:49:42.390         95460.820299         95509.387491  ...   
2022-12-03 06:49:42.400         95455.37366

# Application and Plotting
Lets apply what we learned and make some nice plots.


In [ ]:
sample_start = dt.datetime(2022, 7, 22, 23)
sample_finish = dt.datetime(2022, 7, 22, 23, 4)

In [ ]:
df, data_limit_applied = client.get_sensor_data(
    installation_reference="aventa-turbine-test",
    node_id="2",
    sensor_type_reference="barometer_thermometer",
    start=sample_start,
    finish=sample_finish,
    row_limit=100,
)

data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
signal_df = df[["datetime"]+data_columns].set_index('datetime')
signal_df.columns = sensor_types_metadata["barometer_thermometer"]["sensors"]
thermometer = RawSignal(signal_df, "barometer_thermometer")
thermometer.pad_gaps(dt.timedelta(seconds=1))
thermometer.measurement_to_variable()


df, data_limit_applied = client.get_sensor_data(
    installation_reference="aventa-turbine-test",
    node_id="2",
    sensor_type_reference="barometer",
    start=sample_start,
    finish=sample_finish,
    row_limit=100,
)

data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
signal_df = df[["datetime"]+data_columns].set_index('datetime')
signal_df.columns = sensor_types_metadata["barometer"]["sensors"]
barometer = RawSignal(signal_df, "barometer")
barometer.pad_gaps(dt.timedelta(seconds=1))
barometer.measurement_to_variable()



We can filter out unreasonable temperature readings:

In [ ]:
barometer.dataframe = barometer.dataframe[(barometer.dataframe <= 150e3) & (barometer.dataframe >= 60e3)]

In [ ]:
thermo_filter = ((thermometer.dataframe <= 45) & (thermometer.dataframe >= 5))
thermometer.dataframe = thermometer.dataframe[thermo_filter]

We can filter baro data with thermo filter as it is more obvious from temperature when the data is reasonable.

In [ ]:
thermo_filter.columns=barometer.dataframe.columns 
barometer_processed = RawSignal(barometer.dataframe[thermo_filter], "barometer")

In [ ]:
barometer.plot(sensor_types_metadata)

In [ ]:
thermometer.plot(sensor_types_metadata)

Get atm. pressure in still conditions

In [ ]:
still_sample_start = dt.datetime(2022, 7, 30, 22, 8, 30)
still_sample_finish = dt.datetime(2022, 7, 30, 22, 10)

df, data_limit_applied = client.get_sensor_data(
    installation_reference="aventa-turbine-test",
    node_id="2",
    sensor_type_reference="barometer",
    start=still_sample_start,
    finish=still_sample_finish,
    row_limit=3000,
)

data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
signal_df = df[["datetime"]+data_columns].set_index('datetime')
signal_df.columns = sensor_types_metadata["barometer"]["sensors"]
still_barometer = RawSignal(signal_df, "barometer")
still_barometer.pad_gaps(dt.timedelta(seconds=1))
still_barometer.measurement_to_variable()

In [ ]:
resampled_df = still_barometer.dataframe.resample('1S').mean()

In [ ]:
plots.plot_sensors(resampled_df.reset_index())

In [ ]:
still_barometer.plot(sensor_types_metadata)

In [ ]:
atm_pressure=still_barometer.dataframe.mean()
barometer_processed.dataframe -= atm_pressure
barometer_processed.dataframe.columns=range(40)

In [ ]:
barometer_processed.dataframe -= 96.2e3

In [ ]:
barometer_processed = sensor_signals["barometer"]
barometer_processed.dataframe.columns=range(40)

In [ ]:
df, data_limit_applied = client.get_sensor_data(
    installation_reference="predep-test-20221130",
    node_id="1",
    sensor_type_reference="accelerometer",
    start=sample_start,
    finish=sample_finish,
    row_limit=1000000000,
)

data_columns =  df.columns[df.columns.str.startswith('f')].tolist()
signal_df = df[["datetime"]+data_columns].set_index('datetime')
signal_df.columns = sensor_types_metadata["accelerometer"]["sensors"]
accelerometer = RawSignal(signal_df, "accelerometer")
accelerometer.pad_gaps(dt.timedelta(seconds=1))
accelerometer.measurement_to_variable()

In [ ]:
accelerometer.plot(sensor_types_metadata)

In [ ]:
barometer_coordinates = pd.DataFrame(
{
    "x":[0.0922119084253499,0.0784111428350326,0.0635467289600351,0.0494718121475968,0.0348007242384172,0.0209836333929474,0.00821066059356812,0.000287016245904044,0.00404525690786845,0.0144832434319023,0.0262233800596873,0.0400800665711656,0.0527383343496627,0.067091235212706,0.0806198869961322,0.0950907543239808,0.109385753674316,0.124095983654786,0.137965865257593,0.152444428050515,0.166519849421778,0.180888288065098,0.194997584196934,0.209244523125562,0.205889294568664,0.191642397539166,0.177546273531482,0.16341218630832,0.149101720253823,0.13520156795253,0.120998714181473, 0.106888086108249, 0.317522175046627,0.303212316451663,0.289521584134576,0.27539208898574,0.261557829260995,0.247306755467187,0.233510956423097,0.219404694503124],
    "y":[-0.0298805503559557,-0.0289602925497801,-0.0286682644007734,-0.0282494451123373,-0.0283085310605089,-0.0220526447236176,-0.0150202186794781,-0.00153116363587266,0.0122178780781479,0.0233181734120601,0.0297669723860122,0.0366470139798434,0.0406771795779223,0.0451941341237041,0.0467644784370817,0.0496846161412822,0.0498060778947494,0.0513140914378661,0.0504517682647457,0.050052270897714,0.049185007882442,0.0476147087689916,0.0462310698726144,0.0442390615811808,-0.0255947526614279,-0.0259384688248189,-0.0270997119255558,-0.0275706114026977,-0.0290218042880069,-0.0288830613850707,-0.0292499894120214, -0.0294782194547466, -0.0172704802584763,-0.017977126230662,-0.0191442611821929,-0.0200474513720925,-0.0210341817018639,-0.0218470990312465,-0.0230808054519725,-0.0238661186683629],
    "sensor": barometer_processed.dataframe.columns
 })

In [ ]:
import plotly.express as px

fig = px.scatter(barometer_coordinates, x="x", y="y", text="sensor")
fig.update_traces(textposition='top center')
fig.update_layout({"title":"Sensor Locations"})
fig.show()

In [ ]:
x_pressure_side = barometer_coordinates[barometer_coordinates["y"]<0]["x"].sort_values()
x_suction_side = barometer_coordinates[barometer_coordinates["y"]>0]["x"].sort_values()

suction_side_barometers = barometer_processed.dataframe.columns[x_suction_side.index]
pressure_side_barometers = barometer_processed.dataframe.columns[x_pressure_side.index]

In [ ]:
pressure_side_barometers

In [ ]:
import plotly.graph_objects as go
import plotly.io as pio

pio.renderers.default = "colab"

ani_samples = 100

#ani_start=dt.datetime(2022, 7, 22, 23, 2, 30)
#aero_pressure=aero_pressure[aero_pressure.index>ani_start]

suction_df = barometer_processed.dataframe[suction_side_barometers]
pressure_df = barometer_processed.dataframe[pressure_side_barometers]

#pressure_df = pressure_df.drop(["f19_", "f29_"], axis=1)
#pressure_df = pressure_df.drop(["f2_"], axis=1)

plot_frames = [go.Frame(
    data=[
        go.Scatter(            
            x=x_pressure_side,
            y=pressure_df.iloc[i],
            text=pressure_df.columns,
            textposition='top center',
            name="Pressure side ",
            mode='lines+markers+text',
            marker=dict(color="red")
        ),
        go.Scatter(            
            x=x_suction_side,
            y=suction_df.iloc[i],
            text=suction_df.columns,
            textposition='top center',
            name="Suction side ",
            mode='lines+markers+text',
            marker=dict(color="blue")
        )        
    ],
    name=str(suction_df.index[i])
    )
for i in range(ani_samples)]


sliders_dict = {
    "active": 0,
    "yanchor": "top",
    "xanchor": "left",
    "currentvalue": {
        "font": {"size": 20},
        "prefix": "Time:",
        "visible": True,
        "xanchor": "right"
    },
    "transition": {"duration": 60},
    "pad": {"b": 10, "t": 50},
    "len": 0.9,
    "x": 0.1,
    "y": 0,
    "steps": [{"args": [
        [str(suction_df.index[i])],
        {"frame": {"duration": 60, "redraw": True},
         "mode": "immediate",
         "transition": {"duration": 0}}
    ],
        "label": str(suction_df.index[i]),
        "method": "animate"} for i in range(ani_samples)]
}

fig = go.Figure(
data=plot_frames[0]["data"],
layout=go.Layout(
        title_text="Cp Curve", hovermode="closest",
        xaxis=dict(title="Chordwise sensor postition [m/1m]", range=[0, 0.35], autorange=False),
        yaxis=dict(title="Aerodynamic Pressure [Pa]", range=[2.5e3, -1e3], autorange=False),
        sliders=[sliders_dict],
        updatemenus=[dict(type="buttons",
                          buttons=[dict(label="\u25B6",
                                        method="animate",
                                        args=[None,
                                              {"frame": {"duration": 60, "redraw": True},
                                               "transition": {"duration": 0}}]),
                                   dict(label="Pause",
                                        method="animate",
                                        args=[[None], {"frame": {"duration": 0, "redraw": False},
                                  "mode": "immediate",
                                  "transition": {"duration": 0}}])
                                   ],
                          yanchor="bottom",
                          y=-0.5,
                          )]),
frames = plot_frames

)

fig.show()